### PR TITLE
PE-234 - Add Amazon CloudFront signed URLs endpoint.

### PR DESCRIPTION
**Pull request by Squirrel18**
## Description:

This PR adds a new endpoint to sign URLs related to any resource in Amazon CloudFront service.

## Previous work:

https://github.com/proversity-org/edx-platform/pull/1036

## Configure:

ENV_TOKENS:

- SCV_CLOUDFRONT_ID = 'your-cloudfront-project-id'
- SCV_CLOUDFRONT_URL = 'your-cloudfront-url-containing-the-slash-at-the-end'

AUTH_TOKENS

- SCV_CLOUDFRONT_PRIVATE_SIGNING_KEY = 'cloudfront-private-signing-key'

## Reviewers:

- [ ] @andrey-canon

### CODE REVIEW

File: secure_cloudfront_video/views.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    """
+    Generate a redirect to the AWS CloudFront resource.
+
+    The signed resource URl must point to /secure-cloudfront-video/?key=path-to-aws-resource
+    The resource URL will have a expiration time of one minute.
+    Note that the resource URl must have the slash at the beginning of the string.
+
+    **Example Requests**:
+
+        GET lms-ur/secure-cloudfront-video/?key=path-to-aws-resource
```
Comment (andrey-canon) [inline secure_cloudfront_video/views.py]:
```suggestion
        GET lms-url/secure-cloudfront-video/?key=path-to-aws-resource
```

File: secure_cloudfront_video/utils.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+    More info: https://github.com/boto/botocore/blob/develop/botocore/signers.py#L274
+
+    Returns:
+        String that contains the signed private key.
+    """
+    cloudfront_key = str(getattr(settings, 'SCV_CLOUDFRONT_PRIVATE_SIGNING_KEY', ''))
+
+    if not cloudfront_key:
+        raise Http404
```
Comment (andrey-canon) [inline secure_cloudfront_video/utils.py]: This is not a view and it should not raise an 404, use another exception a handle it on the view

File: secure_cloudfront_video/views.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    key = request.GET.get('key', '')
+
+    if not key:
+        raise Http404
+
+    cloudfront_url = getattr(settings, 'SCV_CLOUDFRONT_URL', '')
+    cloudfront_id = getattr(settings, 'SCV_CLOUDFRONT_ID', '')
+
+    if not (cloudfront_url or cloudfront_id):
```
Comment (andrey-canon) [inline secure_cloudfront_video/views.py]: @Squirrel18 I think this should be an `and` and you could merge this with the previous conditional
somethin llike   `if not (key and cloudfront_url and cloudfront_id):`

File: secure_cloudfront_video/views.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        raise Http404
+
+    resource_url = '{}{}'.format(cloudfront_url, key)
+    cloudfront_signer = CloudFrontSigner(cloudfront_id, cloudfront_rsa_signer)
+    redirect_url = cloudfront_signer.generate_presigned_url(
+        resource_url,
```
Comment (andrey-canon) [inline secure_cloudfront_video/views.py]:
```suggestion
        url='{}{}'.format(cloudfront_url, key),
```

File: secure_cloudfront_video/utils.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    Return the UTC now time plus one minute.
+
+    Returns:
+        datetime.datetime instance that contains the UTC time plus one minute.
+    """
+    minute_delta = timedelta(minutes=1)
+    time_now = datetime.utcnow()
+
+    return time_now + minute_delta
```
Comment (andrey-canon) [inline secure_cloudfront_video/utils.py]: All these lines could be merged

File: secure_cloudfront_video/utils.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    Returns:
+        String that contains the signed private key.
+    """
+    cloudfront_key = str(getattr(settings, 'SCV_CLOUDFRONT_PRIVATE_SIGNING_KEY', ''))
+
+    if not cloudfront_key:
+        raise Http404
+
+    private_key = serialization.load_pem_private_key(
+        cloudfront_key,
```
Comment (andrey-canon) [inline secure_cloudfront_video/utils.py]:
```suggestion
        data=cloudfront_key,
```

**Review by andrey-canon — CHANGES_REQUESTED** [review]

File: secure_cloudfront_video/views.py (hunk 1 above)
Comment (Squirrel18) [reply secure_cloudfront_video/views.py]: Addressed.

File: secure_cloudfront_video/views.py (hunk 3 above)
Comment (Squirrel18) [reply secure_cloudfront_video/views.py]: Addressed.

File: secure_cloudfront_video/utils.py (hunk 5 above)
Comment (Squirrel18) [reply secure_cloudfront_video/utils.py]: Addressed.

File: secure_cloudfront_video/utils.py (hunk 6 above)
Comment (Squirrel18) [reply secure_cloudfront_video/utils.py]: Addressed.

File: secure_cloudfront_video/utils.py (hunk 2 above)
Comment (Squirrel18) [reply secure_cloudfront_video/utils.py]: Addressed.

File: secure_cloudfront_video/views.py (hunk 4 above)
Comment (Squirrel18) [reply secure_cloudfront_video/views.py]: Addressed.

**Review by andrey-canon — APPROVED** [review]

**Pull request merged**